File: Emulator/src/main/core/game/world/map/RegionPlane.java
package core.game.world.map;

import core.game.node.Node;
import core.game.node.entity.npc.NPC;
import core.game.node.entity.player.Player;
import core.game.node.item.GroundItem;
import core.game.node.item.Item;
import core.game.node.scenery.Scenery;
import core.game.world.map.build.DynamicRegion;
import core.game.world.map.build.RegionFlags;
import core.game.world.update.flag.chunk.ItemUpdateFlag;
import core.net.packet.PacketRepository;
import core.net.packet.context.BuildItemContext;
import core.net.packet.out.ClearGroundItem;
import core.net.packet.out.ConstructGroundItem;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Objects;
import java.util.concurrent.CopyOnWriteArrayList;

/**
 * The type Region plane.
 */
public final class RegionPlane {

    /**
     * The constant REGION_SIZE.
     */
    public static final int REGION_SIZE = 64;

    /**
     * The constant CHUNK_SIZE.
     */
    public static final int CHUNK_SIZE = REGION_SIZE >> 3;

    /**
     * The constant NULL_OBJECT.
     */
    public static final Scenery NULL_OBJECT = new Scenery(0, Location.create(0, 0, 0));

	private final int plane;

	private final Region region;

	private final RegionFlags flags;

	private final RegionFlags projectileFlags;

	private final RegionChunk[][] chunks;

	private final List<NPC> npcs;

	private final List<Player> players;

	private Scenery[][] objects;

    /**
     * Instantiates a new Region plane.
     *
     * @param region the region
     * @param plane  the plane
     */
    public RegionPlane(Region region, int plane) {
		this.plane = plane;
		this.region = region;
		this.players = new CopyOnWriteArrayList<Player>();
		this.npcs = new CopyOnWriteArrayList<NPC>();
		Location base = region.getBaseLocation();
		this.flags = new RegionFlags(plane, base.getX(), base.getY());
		this.projectileFlags = new RegionFlags(plane, base.getX(), base.getY(), true);
		this.objects = new Scenery[REGION_SIZE][REGION_SIZE];
		this.chunks = new RegionChunk[CHUNK_SIZE][CHUNK_SIZE];
	}

    /**
     * Pulse.
     */
    public void pulse() {
		Arrays.stream(chunks).forEach(regionChunks -> {Arrays.stream(regionChunks).filter(Objects::nonNull).forEach(RegionChunk::resetFlags);});
	}

    /**
     * Add.
     *
     * @param object    the object
     * @param x         the x
     * @param y         the y
     * @param landscape the landscape
     */
    public void add(Scenery object, int x, int y, boolean landscape) {
		setChunkObject(x, y, object);
		if (landscape) {
			objects[x][y] = object;
		}
		if (object != null) {
			object.setRenderable(true);
		}
	}

    /**
     * Gets region chunk.
     *
     * @param chunkX the chunk x
     * @param chunkY the chunk y
     * @return the region chunk
     */
    public RegionChunk getRegionChunk(int chunkX, int chunkY) {
		RegionChunk r = chunks[chunkX][chunkY];
		if (r != null) {
			return r;
		}
		if (region.isBuild()) {
			return chunks[chunkX][chunkY] = new BuildRegionChunk(region.getBaseLocation().transform(chunkX << 3, chunkY << 3, plane), 0, this);
		}
		return chunks[chunkX][chunkY] = new RegionChunk(region.getBaseLocation().transform(chunkX << 3, chunkY << 3, plane), 0, this);
	}

    /**
     * Sets region chunk.
     *
     * @param chunkX the chunk x
     * @param chunkY the chunk y
     * @param chunk  the chunk
     */
    public void setRegionChunk(int chunkX, int chunkY, RegionChunk chunk) {
		chunks[chunkX][chunkY] = chunk;
	}

    /**
     * Remove.
     *
     * @param x the x
     * @param y the y
     */
    public void remove(int x, int y) {
		remove(x, y, -1);
	}

    /**
     * Remove.
     *
     * @param x        the x
     * @param y        the y
     * @param objectId the object id
     */
    public void remove(int x, int y, int objectId) {
		int chunkX = x / CHUNK_SIZE;
		int chunkY = y / CHUNK_SIZE;
		int offsetX = x - chunkX * CHUNK_SIZE;
		int offsetY = y - chunkY * CHUNK_SIZE;
		RegionChunk chunk = getRegionChunk(chunkX, chunkY);
		Scenery remove = new Scenery(0, region.getBaseLocation().transform(x, y, plane), 22, 0);
		remove.setRenderable(false);
		if (chunk instanceof BuildRegionChunk) {
			int index = ((BuildRegionChunk) chunk).getIndex(offsetX, offsetY, objectId);
			((BuildRegionChunk) chunk).getObjects(index)[offsetX][offsetY] = remove;
			return;
		}
		chunk.getObjects()[offsetX][offsetY] = remove;
	}

	private void setChunkObject(int x, int y, Scenery object) {
		int chunkX = x / CHUNK_SIZE;
		int chunkY = y / CHUNK_SIZE;
		int offsetX = x - chunkX * CHUNK_SIZE;
		int offsetY = y - chunkY * CHUNK_SIZE;
		RegionChunk r = getRegionChunk(chunkX, chunkY);
		if (r instanceof BuildRegionChunk) {
			((BuildRegionChunk) r).store(object);
			return;
		}
		r.getObjects()[offsetX][offsetY] = object;
	}

    /**
     * Get objects scenery [ ] [ ].
     *
     * @return the scenery [ ] [ ]
     */
    public Scenery[][] getObjects() {
		return objects;
	}

    /**
     * Gets object list.
     *
     * @return the object list
     */
    public List<Scenery> getObjectList() {
            ArrayList<Scenery> list = new ArrayList();
            for (int x = 0; x < REGION_SIZE; x++) {
                for (int y = 0; y < REGION_SIZE; y++) {
                    if (objects[x][y] != null)
                        list.add(objects[x][y]);
                }
            }
            return list;
        }

    /**
     * Clear.
     */
    public void clear() {
		for (RegionChunk[] c : chunks) {
			for (RegionChunk chunk : c) {
				if (chunk != null) {
					chunk.clear();
				}
			}
		}
		if (region instanceof DynamicRegion && objects != null) {
			for (int x = 0; x < REGION_SIZE; x++) {
				for (int y = 0; y < REGION_SIZE; y++) {
					objects[x][y] = null;
				}
			}
			objects = null;
		}
	}

    /**
     * Add.
     *
     * @param npc the npc
     */
    public void add(NPC npc) {
		npcs.add(npc);
	}

    /**
     * Add.
     *
     * @param player the player
     */
    public void add(Player player) {
		players.add(player);
	}

    /**
     * Add.
     *
     * @param item the item
     */
    public void add(GroundItem item) {
		Location l = item.getLocation();
		RegionChunk c = getRegionChunk(l.getLocalX() / RegionChunk.SIZE, l.getLocalY() / RegionChunk.SIZE);
		if (!c.getItems().add(item)) {
			return;
		}
		GroundItem g = (GroundItem) item;
		if (g.isPrivate()) {
			if (g.getDropper() != null) {
				PacketRepository.send(ConstructGroundItem.class, new BuildItemContext(g.getDropper(), item));
			}
			return;
		}
		c.flag(new ItemUpdateFlag(g, ItemUpdateFlag.CONSTRUCT_TYPE));
	}

    /**
     * Remove.
     *
     * @param npc the npc
     */
    public void remove(NPC npc) {
		npcs.remove(npc);
	}

    /**
     * Remove.
     *
     * @param player the player
     */
    public void remove(Player player) {
		players.remove(player);
	}

    /**
     * Remove.
     *
     * @param item the item
     */
    public void remove(GroundItem item) {
		Location l = item.getLocation();
		RegionChunk c = getRegionChunk(l.getLocalX() / RegionChunk.SIZE, l.getLocalY() / RegionChunk.SIZE);
		if (!c.getItems().remove(item)) {
			return;
		}
		if (item.isPrivate()) {
			if (item.getDropper() != null && item.getDropper().isPlaying() && item.getDropper().getLocation().withinDistance(l)) {
				PacketRepository.send(ClearGroundItem.class, new BuildItemContext(item.getDropper(), item));
			}
			return;
		}
		c.flag(new ItemUpdateFlag(item, ItemUpdateFlag.REMOVE_TYPE));
	}

    /**
     * Gets flags.
     *
     * @return the flags
     */
    public RegionFlags getFlags() {
		return flags;
	}

    /**
     * Gets projectile flags.
     *
     * @return the projectile flags
     */
    public RegionFlags getProjectileFlags() {
		return projectileFlags;
	}

    /**
     * Gets npcs.
     *
     * @return the npcs
     */
    public List<NPC> getNpcs() {
		return npcs;
	}

    /**
     * Gets entities.
     *
     * @return the entities
     */
    public List<Node> getEntities()
	{
	    List<Node> entities = new ArrayList<>(npcs);
		Arrays.stream(getObjects()).forEach(o -> entities.addAll(Arrays.asList(o)));
	    return  entities;
	}

    /**
     * Gets players.
     *
     * @return the players
     */
    public List<Player> getPlayers() {
		return players;
	}

    /**
     * Gets plane.
     *
     * @return the plane
     */
    public int getPlane() {
		return plane;
	}

    /**
     * Gets region.
     *
     * @return the region
     */
    public Region getRegion() {
		return region;
	}

    /**
     * Gets chunk object.
     *
     * @param x the x
     * @param y the y
     * @return the chunk object
     */
    public Scenery getChunkObject(int x, int y) {
		return getChunkObject(x, y, -1);
	}

    /**
     * Gets chunk object.
     *
     * @param x        the x
     * @param y        the y
     * @param objectId the object id
     * @return the chunk object
     */
    public Scenery getChunkObject(int x, int y, int objectId) {
		int chunkX = x / CHUNK_SIZE;
		int chunkY = y / CHUNK_SIZE;
		int offsetX = x - chunkX * CHUNK_SIZE;
		int offsetY = y - chunkY * CHUNK_SIZE;
		RegionChunk chunk = getRegionChunk(chunkX, chunkY);
		if (chunk instanceof BuildRegionChunk) {
			BuildRegionChunk brc = (BuildRegionChunk) chunk;
			return brc.get(offsetX, offsetY, brc.getIndex(offsetX, offsetY, objectId));
		}
		return getRegionChunk(chunkX, chunkY).getObjects()[offsetX][offsetY];
	}

    /**
     * Gets chunk items.
     *
     * @param x the x
     * @param y the y
     * @return the chunk items
     */
    public List<GroundItem> getChunkItems(int x, int y) {
		int chunkX = x / CHUNK_SIZE;
		int chunkY = y / CHUNK_SIZE;
		return getRegionChunk(chunkX, chunkY).getItems();
	}

    /**
     * Gets item.
     *
     * @param itemId the item id
     * @param l      the l
     * @param player the player
     * @return the item
     */
    public GroundItem getItem(int itemId, Location l, Player player) {
		GroundItem groundItem = null;
		for (Item item : getChunkItems(l.getLocalX(), l.getLocalY())) {
			GroundItem g = (GroundItem) item;
			if (g.getId() == itemId && l.equals(g.getLocation()) && !g.isRemoved()) {
				if (groundItem != null && (!g.isPrivate() || player == null)) {
					continue;
				}
				if ((!g.isPrivate() || player == null || g.droppedBy(player))) {
					groundItem = g;
				}
			}
		}
		return groundItem;
	}

    /**
     * Get chunks region chunk [ ] [ ].
     *
     * @return the region chunk [ ] [ ]
     */
    public RegionChunk[][] getChunks() {
		return chunks;
	}

}

File: Emulator/src/main/core/game/world/map/RegionPlane.kt
package core.game.world.map

import core.game.node.Node
import core.game.node.entity.npc.NPC
import core.game.node.entity.player.Player
import core.game.node.item.GroundItem
import core.game.node.item.Item
import core.game.node.scenery.Scenery
import core.game.world.map.build.DynamicRegion
import core.game.world.map.build.RegionFlags
import core.game.world.update.flag.chunk.ItemUpdateFlag
import core.net.packet.OutgoingContext
import core.net.packet.PacketRepository
import core.net.packet.out.ClearGroundItem
import core.net.packet.out.ConstructGroundItem
import java.util.concurrent.CopyOnWriteArrayList

/**
 * Represents a single plane (z-level) of a region, managing its tiles, objects, entities, and flags.
 *
 * @property region The parent region this plane belongs to.
 * @property plane The plane index (0–3) representing the height level.
 */
class RegionPlane(val region: Region, val plane: Int) {

    companion object {
        /**
         * Represents the size of a region in tiles.
         */
        const val REGION_SIZE = 64

        /**
         * Represents the size of a chunk in tiles (8x8).
         */
        const val CHUNK_SIZE = REGION_SIZE shr 3

        /**
         * Represents the object used when an object is removed or not found.
         */
        val NULL_OBJECT = Scenery(0, Location.create(0, 0, 0))
    }

    /**
     * Represents the clipping and movement flags for the plane.
     */
    val flags = RegionFlags(plane, region.baseLocation.x, region.baseLocation.y, false)

    /**
     * Represents the projectile clipping flags.
     */
    val projectileFlags = RegionFlags(plane, region.baseLocation.x, region.baseLocation.y, true)

    /**
     * Represents the 2D array of chunks (8x8) for this plane.
     */
    val chunks = Array(CHUNK_SIZE) { arrayOfNulls<RegionChunk>(CHUNK_SIZE) }

    /**
     * Represents the list of NPCs currently on this plane.
     */
    val npcs: MutableList<NPC> = CopyOnWriteArrayList()

    /**
     * Represents the list of players currently on this plane.
     */
    val players: MutableList<Player> = CopyOnWriteArrayList()

    /**
     * Represents the 2D array representing placed scenery objects.
     */
    var objects: Array<Array<Scenery?>>? = Array(REGION_SIZE) { arrayOfNulls<Scenery>(REGION_SIZE) }

    /**
     * Resets all chunk flags within this plane.
     */
    fun pulse() {
        chunks.flatten().filterNotNull().forEach { it.resetFlags() }
    }

    /**
     * Adds a scenery object to the world and marks it renderable.
     *
     * @param obj The scenery object to place.
     * @param x The x-coordinate (local to the region).
     * @param y The y-coordinate (local to the region).
     * @param landscape Whether to store the object in the region's landscape object array.
     */
    fun add(obj: Scenery?, x: Int, y: Int, landscape: Boolean) {
        setChunkObject(x, y, obj)
        if (landscape) objects?.get(x)?.set(y, obj)
        obj?.isRenderable = true
    }

    /**
     * Returns or initializes the chunk at the given chunk coordinates.
     */
    fun getRegionChunk(chunkX: Int, chunkY: Int): RegionChunk {
        return chunks[chunkX][chunkY] ?: run {
            val location = region.baseLocation.transform(chunkX shl 3, chunkY shl 3, plane)
            val chunk = if (region.build) BuildRegionChunk(location, 0, this) else RegionChunk(location, 0, this)
            chunks[chunkX][chunkY] = chunk
            chunk
        }
    }

    /**
     * Sets a specific chunk manually in the plane.
     */
    fun setRegionChunk(chunkX: Int, chunkY: Int, chunk: RegionChunk) {
        chunks[chunkX][chunkY] = chunk
    }

    /**
     * Removes a scenery object at the given coordinates.
     *
     * @param x The x-coordinate (local to region).
     * @param y The y-coordinate (local to region).
     * @param objectId Optional object ID to help locate the object in a dynamic chunk.
     */
    fun remove(x: Int, y: Int, objectId: Int = -1) {
        val chunkX = x / CHUNK_SIZE
        val chunkY = y / CHUNK_SIZE
        val offsetX = x % CHUNK_SIZE
        val offsetY = y % CHUNK_SIZE
        val chunk = getRegionChunk(chunkX, chunkY)
        val removed = Scenery(0, region.baseLocation.transform(x, y, plane), 22, 0).apply { isRenderable = false }
        if (chunk is BuildRegionChunk) {
            val index = chunk.getIndex(offsetX, offsetY, objectId)
            chunk.getObjects(index)[offsetX][offsetY] = removed
        } else {
            chunk.objects[offsetX][offsetY] = removed
        }
    }

    /**
     * Stores an object in the correct chunk at the specified tile.
     */
    private fun setChunkObject(x: Int, y: Int, obj: Scenery?) {
        val chunkX = x / CHUNK_SIZE
        val chunkY = y / CHUNK_SIZE
        val offsetX = x % CHUNK_SIZE
        val offsetY = y % CHUNK_SIZE
        val chunk = getRegionChunk(chunkX, chunkY)
        if (chunk is BuildRegionChunk) {
            chunk.store(obj)
        } else {
            chunk.objects[offsetX][offsetY] = obj
        }
    }

    /**
     * Returns a flat list of all scenery objects present in this plane.
     */
    fun getObjectList(): List<Scenery> = buildList {
        objects?.forEach { row -> row.filterNotNull().forEach { add(it) } }
    }

    /**
     * Clears all data from this plane, including chunks and objects.
     */
    fun clear() {
        chunks.flatten().filterNotNull().forEach { it.clear() }
        if (region is DynamicRegion && objects != null) {
            for (x in 0 until REGION_SIZE) {
                for (y in 0 until REGION_SIZE) {
                    objects!![x][y] = null
                }
            }
            objects = null
        }
    }

    /**
     * Adds an NPC to the plane.
     */
    fun add(npc: NPC) { npcs.add(npc) }

    /**
     * Adds a player to the plane.
     */
    fun add(player: Player) { players.add(player) }

    /**
     * Adds a ground item to the appropriate chunk.
     */
    fun add(item: GroundItem) {
        val l = item.location
        val chunk = getRegionChunk(l.localX / RegionChunk.SIZE, l.localY / RegionChunk.SIZE)
        if (!chunk.items.add(item)) return
        if (item.isPrivate) {
            item.dropper?.let { PacketRepository.send(ConstructGroundItem::class.java, OutgoingContext.BuildItem(it, item, 0)) }
        } else {
            chunk.flag(ItemUpdateFlag(item, ItemUpdateFlag.CONSTRUCT_TYPE))
        }
    }

    /** Removes an NPC from the plane. */
    fun remove(npc: NPC) { npcs.remove(npc) }

    /** Removes a player from the plane. */
    fun remove(player: Player) { players.remove(player) }

    /**
     * Removes a ground item from the appropriate chunk.
     */
    fun remove(item: GroundItem) {
        val l = item.location
        val chunk = getRegionChunk(l.localX / RegionChunk.SIZE, l.localY / RegionChunk.SIZE)
        if (!chunk.items.remove(item)) return
        if (item.isPrivate) {
            val dropper = item.dropper
            if (dropper != null && dropper.isPlaying && dropper.location.withinDistance(l)) {
                PacketRepository.send(ClearGroundItem::class.java, OutgoingContext.BuildItem(dropper, item, 0))
            }
        } else {
            chunk.flag(ItemUpdateFlag(item, ItemUpdateFlag.REMOVE_TYPE))
        }
    }

    /**
     * Returns all items within a specific chunk.
     */
    fun getChunkItems(x: Int, y: Int): List<GroundItem> {
        val chunkX = x / CHUNK_SIZE
        val chunkY = y / CHUNK_SIZE
        return getRegionChunk(chunkX, chunkY).items
    }

    /**
     * Retrieves a specific ground item by ID and location, with optional visibility check for a player.
     *
     * @param itemId The ID of the item.
     * @param l The location of the item.
     * @param player Optional player to check private ownership.
     */
    fun getItem(itemId: Int, l: Location, player: Player?): GroundItem? {
        var match: GroundItem? = null
        for (item in getChunkItems(l.localX, l.localY)) {
            val g = item as GroundItem
            if (g.id == itemId && l == g.location && !g.isRemoved) {
                if (match != null && (!g.isPrivate || player == null)) continue
                if (!g.isPrivate || player == null || g.droppedBy(player)) {
                    match = g
                }
            }
        }
        return match
    }

    /**
     * Retrieves an object from the plane at the given tile.
     *
     * @param x The x-coordinate (local to region).
     * @param y The y-coordinate (local to region).
     * @param objectId Optional object ID for dynamic chunk lookup.
     */
    fun getChunkObject(x: Int, y: Int, objectId: Int = -1): Scenery? {
        val chunkX = x / CHUNK_SIZE
        val chunkY = y / CHUNK_SIZE
        val offsetX = x % CHUNK_SIZE
        val offsetY = y % CHUNK_SIZE
        val chunk = getRegionChunk(chunkX, chunkY)
        return if (chunk is BuildRegionChunk) {
            val index = chunk.getIndex(offsetX, offsetY, objectId)
            chunk.get(offsetX, offsetY, index)
        } else {
            chunk.objects[offsetX][offsetY]
        }
    }

    /**
     * Returns a list of all entity nodes (NPCs and scenery) in this plane.
     */
    fun getEntities(): List<Node> = buildList {
        addAll(npcs)
        objects?.forEach { row -> row.filterNotNull().forEach { add(it) } }
    }
}